fabric-input-events-v1/src/main/java/net/fabricmc/fabric/impl/client/input/InputCallbacksImpl.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.client.input;
+
+import java.util.Map;
+
+import org.lwjgl.glfw.GLFW;
+import org.lwjgl.glfw.GLFWDropCallback;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+
+import net.minecraft.client.options.KeyBinding;
+import net.minecraft.client.util.InputUtil;
+import net.minecraft.client.util.InputUtil.Key;
+
+import net.fabricmc.fabric.api.event.client.input.ClientInputEvents;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.InputUtilTypeAccessor;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.KeyBindingAccessor;
+
+public final class InputCallbacksImpl {
+	public static void onKey(long window, int code, int scancode, int action, int modKeys) {
+		FabricKeyboardImpl.updateModKeys(modKeys);
+
+		ClientInputEvents.KEY.invoker().onKeyChanged(code, scancode, action, modKeys);
+
+		boolean hasListeners = ClientInputEvents.KEY_PRESSED.hasListeners()
+							|| ClientInputEvents.KEY_RELEASED.hasListeners()
+							|| ClientInputEvents.KEY_REPEATED.hasListeners()
+							|| ClientInputEvents.KEYBIND_PRESSED.hasListeners()
+							|| ClientInputEvents.KEYBIND_RELEASED.hasListeners()
+							|| ClientInputEvents.KEYBIND_REPEATED.hasListeners();
+
+		if (hasListeners) {
+			Map<InputUtil.Key, KeyBinding> keyToBindings = KeyBindingAccessor.getKeyToBindings();
+			Key key = InputUtil.fromKeyCode(code, scancode);
+			KeyBinding binding = keyToBindings.get(key);
```

Comment:
You should move this to some sort of cached supplier, which you can then expose to your event callbacks, instead of adding a `hasListeners` in the events.

Reply:
```java
Lazy<KeyBinding> lazyBinding = new Lazy<>(() -> {
    Map<InputUtil.Key, KeyBinding> keyToBindings = KeyBindingAccessor.getKeyToBindings();
    return keyToBindings.get(InputUtil.fromKeyCode(code, scancode));
});
```
Hmm, I now wonder if you should rather pass in like a functional interface with a method `KeyBinding getKey(int scancode, int keycode)` for performance reasons. It's debatable, but not of priority.

Reply:
This seems to add a lot of complexity and inconvenience, and an extra object allocation every key-press which I was told wasn't allowed, for very little benefit. If we are going to allocate a functional interface anyway, I would much rather go back to the original `KeyEvent` implementation, considering all the benefits I have outlined for that approach earlier.